mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessageZipper.java

```diff
@@ -42,20 +44,37 @@
     }
 
     void zip(Stream<DeletedMessageWithContent> deletedMessages, OutputStream outputStream) throws IOException {
-        ZipArchiveOutputStream zipOutputStream = new ZipArchiveOutputStream(outputStream);
+        try (ZipArchiveOutputStream zipOutputStream = new ZipArchiveOutputStream(outputStream)) {
 
-        deletedMessages
-            .forEach(Throwing.<DeletedMessageWithContent>consumer(
-                deletedMessage -> putMessageToEntry(zipOutputStream, deletedMessage)).sneakyThrow());
+            Optional<IOException> noFailure = Optional.empty();
+            Optional<IOException> failureEncountered = deletedMessages
+                .reduce(noFailure,
+                    (previousFailure, message) -> handleMessage(previousFailure, zipOutputStream, message),
+                    OptionalUtils::or);
```

Comment:
why do you keep only the last failure instead of accumulate them?

Reply:
The `OptionalUtils::or` actually be `OptionalUtils.or(previousError, currentError)`. I accumulate all result, but by pair of previous error & current error.

there is only one exception for the whole process, If in the middle of the stream, I get an exception, then I stop to consume next DeletedMessagesWithContent**s**, but just closing them. Because you said, we don't know the state of ZipOutputStream, thus I i can't handle more deleted messages once I get error.

Reply:
> why do you keep only the last failure instead of accumulate them?

No "fold" in Java 8 ?

Reply:
I understand your strategy @trantienduchn and it is the right thing to do.
But if I can't understand that the code is doing that by reading it, it's probably not written in a simple enough way.
Could you try to make it simpler?

Reply:
I simplified the logic by using traditional forEach, and added more test to cover failling cases. Can you have a look :) ? 72184f6